cryptoprotocol/src/main/java/org/cryptimeleon/incentivesystem/cryptoprotocol/proof/SpendDeductZkp.java

```diff
@@ -59,66 +57,69 @@ public SpendDeductZkp(IncentivePublicParameters incentivePublicParameters, Provi
         var tVar = builder.addZnVariable("t", zn);
         var zStarVar = builder.addZnVariable("zStar", zn);
         var tStarVar = builder.addZnVariable("tStar", zn);
-        var uStarVar = builder.addZnVariable("uStar", zn);
         var uStarInverseVar = builder.addZnVariable("uStarInverse", zn);
-        var eskDecVarVector = new SchnorrZnVariable[pp.getNumEskDigits()];
-        var rVector = new SchnorrZnVariable[pp.getNumEskDigits()];
+        var eskDecVarVector = new ExponentExpressionVector();
+        var rVector = new ExponentExpressionVector();
         for (int i = 0; i < pp.getNumEskDigits(); i++) {
-            eskDecVarVector[i] = builder.addZnVariable("eskStarUserDec_" + i, zn);
-            rVector[i] = builder.addZnVariable("r_" + i, zn);
+            eskDecVarVector = eskDecVarVector.append(builder.addZnVariable("eskStarUserDec_" + i, zn));
+            rVector = rVector.append(builder.addZnVariable("r_" + i, zn));
         }
```

Comment:
This may be very much premature optimization, but every append() causes a copy of the whole vector, so this loop has quadratic runtime.

Maybe do 
```java
var eskDecVarVector = ExponentExpressionVector.generate(i -> builder.addZnVariable("eskStarUserDec_" + i, zn), pp.getNumEskDigits);
var rVector = ExponentExpressionVector.generate(i -> builder.addZnVariable("r_" + i, zn), pp.getNumEskDigits);
```
instead? :)

Reply:
Good idea!